docs/book/ssh-and-sudo-authorization.md

```diff
@@ -256,85 +263,137 @@ able to login.
+their rights should be removed.
+
+Let's mock the current state of this simple ticketing system's API with some data.
+
+```shell
+curl -X PUT localhost:8181/v1/data/elevate -d \
+'{
+  "tickets": {
+    "frontend-dev": "1234"
+  }
+}'
 ```
+This means that for now, if the `frontend-dev` user can provide ticket number `1234`,
+they should be able to SSH into all servers.
+
+Let's write policy to ensure that this happens.
+
+First, we need to make the PAM module take input from the user.
+```shell
+cat >display.rego <<EOF
+package display
+
+# What should be prompted to the user?
+display_spec = [
+  {
+    "message": "Please enter an elevation ticket if you have one:",
+    "style": "prompt_echo_on",
+    "key": "ticket"
+  }
+]
 
-### 6. Elevate a user's rights through policy.
+EOF
+```
+Load this policy into OPA.
 
-Suppose a user needs to be temporarily granted extra privileges. E.g., the user
-may need to debug an issue on a specific production host. In this case, we can
-add an another policy to temporarily allow this.
+```shell
+curl -X PUT --data-binary @display.rego \
+  localhost:8181/v1/policies/display
+```
 
-Let's allow the `frontend-dev` user to run sudo commands on the `frontend` host.
+Then we need to make sure that the authorization takes this input into account.
 
 ```shell
-cat >sudo_authz_special.rego <<EOF
+cat >sudo_authz_elevated.rego <<EOF
+# A package can be defined across multiple files.
 package sudo.authz
 
+import data.elevate
+import input.sysinfo
+import input.display_responses
+
+# Allow this user if the elevation ticket they provided matches our mock API
+# of an internal elevation system.
 allow {
-  input.host_identity.host_id = "frontend"
-  input.user = "frontend-dev"
+  elevate.tickets[sysinfo.pam_username] = display_responses.ticket
 }
 EOF
 ```
 
 Load this policy into OPA.
 
 ```shell
-curl -X PUT --data-binary @sudo_authz_special.rego \
-  localhost:8181/v1/policies/sudo_authz_special
+curl -X PUT --data-binary @sudo_authz_elevated.rego \
+  localhost:8181/v1/policies/sudo_authz_elevated
 ```
 
-Confirm that the user `frontend-dev` can login to the `frontend` host and run sudo commands.
+Confirm that the user `frontend-dev` can indeed use `sudo`.
 
 ```shell
 ssh -p 2222 frontend-dev@localhost \
-  -i ssh_authz.key -o StrictHostKeyChecking=no
+  -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null
+
 sudo ls /
-exit
 ```
+You should be prompted with the message that we defined in our *display* policy
+for both the SSH and `sudo` authorization cycles.
+This happens because the *display* policy is shared by the PAM configurations of SSH and `sudo`.
+In production, it is more practical to use separate policy packages for each PAM configuration.
 
-Lastly, remove the policy from OPA and confirm the user's original rights are restored.
 
-```shell
-curl -i -X DELETE localhost:8181/v1/policies/sudo_authz_special
-```
+For SSH, the value entered does not matter.
```

Comment:
Would rewrite: "The SSH PAM policy does not rely on the input ticket number, so the value does not affect the authorization decision"

Reply:
Done.